torch/testing/_internal/inputgen/variable/space.py

```diff
@@ -0,0 +1,233 @@
+        self.upper = upper
+        self.lower_open = lower_open
+        self.upper_open = upper_open
+
+    def __str__(self) -> str:
+        lower_braket = "(" if self.lower_open else "["
+        upper_braket = ")" if self.upper_open else "]"
+        return f"{lower_braket}{self.lower}, {self.upper}{upper_braket}"
+
+    def empty(self) -> bool:
+        if self.lower < self.upper:
+            return False
+        elif self.lower == self.upper:
+            return self.lower_open or self.upper_open
+        else:
+            return True
+
+    def contains(self, v) -> bool:
+        if self.empty():
+            return False
+        if v < self.lower:
+            return False
+        if v == self.lower and self.lower_open:
+            return False
+        if v == self.upper and self.upper_open:
+            return False
+        if v > self.upper:
+            return False
+        return True
+
+    def contains_int(self) -> bool:
+        if self.empty():
+            return False
+        if self.lower > INT64_MAX or (self.lower == INT64_MAX and self.lower_open):
+            return False
+        if self.upper < INT64_MIN or (self.upper == INT64_MIN and self.upper_open):
+            return False
+        if self.upper - self.lower > 1:
+            return True
+        return self.contains(math.ceil(self.lower)) or self.contains(
+            math.floor(self.upper)
+        )
+
+
+class Intervals:
+    def __init__(self, intervals: Optional[List[Interval]] = None):
+        self.intervals = intervals
+        if self.intervals is None:
+            self.intervals = [Interval()]
+
+    def __str__(self) -> str:
+        if len(self.intervals) == 0:
+            return "{}"
+        return " ".join([str(r) for r in self.intervals])
+
+    def empty(self) -> bool:
+        return all(r.empty() for r in self.intervals)
+
+    def contains(self, v) -> bool:
+        return any(r.contains(v) for r in self.intervals)
+
+    def contains_int(self) -> bool:
+        return any(r.contains_int() for r in self.intervals)
+
+    def remove(self, v) -> None:
+        for ix, r in enumerate(self.intervals):
+            if r.lower <= v <= r.upper:
+                if r.lower == v and r.upper == v:
+                    self.intervals = self.intervals[:ix] + self.intervals[ix + 1 :]
+                elif r.lower == v:
+                    r.lower_open = True
+                elif r.upper == v:
+                    r.upper_open = True
+                else:
+                    new_intervals = [
+                        Interval(r.lower, v, r.lower_open, True),
+                        Interval(v, r.upper, True, r.upper_open),
+                    ]
+                    self.intervals = (
+                        self.intervals[:ix] + new_intervals + self.intervals[ix + 1 :]
+                    )
+                return
+
+    def set_lower(self, lower: float, lower_open: bool = False) -> None:
+        for ix, r in enumerate(self.intervals):
```

Comment:
This seems like a O(N) solution. If you keep your interval sorted it's going to be O(logN)

Reply:
They are sorted. You are right that I could do binary search here. However, for the present use case, the Intervals objects will contain at most 2 or 3 intervals. I don't even know if binary search would be more efficient with so little members, and has the downside of potentially introducing bugs.
True, for any generic Intervals object this is inefficient. But IMO there is no practical need for complicating things at this stage.